api/kv_v2.go

```diff
@@ -551,3 +625,69 @@ func readThenWrite(ctx context.Context, client *Client, mountPath string, secret
 
 	return updatedSecret, nil
 }
+
+func setMetadataMap(metadata KVMetadataInput) (map[string]interface{}, error) {
+	md := make(map[string]interface{})
+	casRequiredKey := "cas_required"
+	deleteVersionAfterKey := "delete_version_after"
+	maxVersionsKey := "max_versions"
+	customMetadataKey := "custom_metadata"
+
+	// --[Explicit Zero Value For The User, Explicit Null On The Backend]--
+	// The KVMetadataInput struct is designed to have pointer fields so that
+	// the user can easily express the desire to explicitly set a field back to its zero
+	// value (e.g. false), as opposed to just having the field remain
+	// unchanged (e.g. nil).
+	// However, in the actual JSONMergePatch request performed by Logical,
+	// the meanings are reversed, where a null value means to delete the
+	// value at that field, and a zero value means to do nothing.
+	//
+	// While cognitively confusing for us on the dev side, this reversal is
+	// by design with users in mind, as we don't want users to accidentally
+	// reset fields just because they neglected to give those fields a value in
+	// the passed KVMetadataInput struct. This way, they only need to pass
+	// the fields they want to change.
+	//
+	// Thus, here we need to dereference our pointers and set any explicit zero values
+	// to null in the JSON so that JSONMergePatch understands to delete the values.
+	if metadata.MaxVersions != nil {
+		derefVal := *(metadata.MaxVersions)
+		if derefVal == 0 {
+			md[maxVersionsKey] = nil
+		} else {
+			md[maxVersionsKey] = derefVal
+		}
+	}
+	if metadata.CASRequired != nil {
+		derefVal := *(metadata.CASRequired)
+		if derefVal == false {
+			md[casRequiredKey] = nil
+		} else {
+			md[casRequiredKey] = derefVal
+		}
+	}
+	if metadata.CustomMetadata != nil {
+		if len(metadata.CustomMetadata) == 0 { // empty non-nil map means delete all the keys
+			md[customMetadataKey] = nil
+		} else {
+			md[customMetadataKey] = metadata.CustomMetadata
+		}
+	}
+	if metadata.DeleteVersionAfter != nil {
+		derefVal := *metadata.DeleteVersionAfter
+		strDuration := derefVal.String()
+		parsedVal, err := time.ParseDuration(strDuration)
+		if err != nil {
+			return nil, fmt.Errorf("unable to parse DeleteVersionAfter duration: %w", err)
+		}
+		if parsedVal == 0 {
+			// TODO: This doesn't work until a bug is fixed.. currently
+			// there seems to be no way to reset the delete_version_after field.
```

Comment:
Is the bug on the server side?

Reply:
This is in fact a bug that exists in the KV secret engine. It is ultimately due to the underlying [durationpb](https://pkg.go.dev/google.golang.org/protobuf/types/known/durationpb#Duration) type that the `delete_version_after` field uses. The struct members have `omitempty` json tags. The JSON patch library that Vault uses accepts byte arrays of marshaled JSON. A duration of "0s" effectively gets marshaled to an empty map, see [this example](https://go.dev/play/p/dEUyjmHdm2h). I have identified a potential fix but would like to chat with some folks before merging it.

Reply:
This has been fixed! https://github.com/hashicorp/vault/pull/15792